### ISSUE TITLE
authCredentials set on DatastoreOptions.builder() is not used so can't access Datastore over http proxy

### ISSUE THREAD

**Issue by successhawk** [GoogleCloudPlatform/google-cloud-java#1210]
Can't use Datastore over an http proxy.  I first reported this under [issue 1138](https://github.com/GoogleCloudPlatform/gcloud-java/issues/1138)
This issue was resolved, however I still can't use Datastore over http proxy due to auth calls not using proxy.  I attempted to configure the AuthCredentials to use my http proxy with the following code:

```
        return DatastoreOptions.builder()
                            .authCredentials(
                                    new MyApplicationDefaultAuthCredentials(
                                            GoogleCredentials.getApplicationDefault(GcpClientUtil.getHttpTransport())))
                            .httpTransportFactory(new CustomHttpTransportFactoryWithProxy())
                            .build()
                            .service());
```

Where MyApplicationDefaultAuthCredentials

```
  /**
   * This was copied and modified so that we could provide an instance of GoogleCredentials
   * 
   * Represents Application Default Credentials, which are credentials that are inferred from the
   * runtime environment.
   *
   * @see <a
   *     href="https://developers.google.com/identity/protocols/application-default-credentials">
   *     Google Application Default Credentials</a>
   */
  static class MyApplicationDefaultAuthCredentials extends AuthCredentials {

    private GoogleCredentials googleCredentials;

    private static final ApplicationDefaultAuthCredentialsState STATE =
        new ApplicationDefaultAuthCredentialsState();

    private static class ApplicationDefaultAuthCredentialsState
        implements RestorableState<AuthCredentials>, Serializable {

      private static final long serialVersionUID = -8839085552021212257L;

      @Override
      public AuthCredentials restore() {
        try {
          return new MyApplicationDefaultAuthCredentials(
                  GoogleCredentials.getApplicationDefault(GcpClientUtil.getHttpTransport()));
        } catch (IOException e) {
          throw new IllegalStateException(
              "Could not restore " + ApplicationDefaultAuthCredentials.class.getSimpleName(), e);
        }
      }

      @Override
      public int hashCode() {
        return getClass().getName().hashCode();
      }

      @Override
      public boolean equals(Object obj) {
        return obj instanceof ApplicationDefaultAuthCredentialsState;
      }
    }

    public MyApplicationDefaultAuthCredentials(GoogleCredentials credentials) {
      googleCredentials = credentials;
    }

    @Override
    public GoogleCredentials credentials() {
      return googleCredentials;
    }

    @Override
    public RestorableState<AuthCredentials> capture() {
      return STATE;
    }
  }
```

Where getHttpTransport is something like follows:

```
public static HttpTransport getHttpTransport() {
    URI uri = getGcpHttpProxyURI();
    if ( uri !=null ) {
        return new NetHttpTransport.Builder().setProxy(new Proxy(Proxy.Type.HTTP,
                new InetSocketAddress(parseHostFromUri(uri),parsePortFromUri(uri)))).build();
    } else {
        return Utils.getDefaultTransport();
    }
}
```

**Comment by mziccard** [comment]
Hi @successhawk. This is strange. I tried the following code:
```java
static final class MyHttpTransportFactory implements HttpServiceOptions.HttpTransportFactory {

  private static final HttpTransport DEFAULT_TRANSPORT = new NetHttpTransport.Builder()
      .setProxy(new Proxy(Proxy.Type.HTTP, new InetSocketAddress("localhost", 3128)))
      .build();
  static final MyHttpTransportFactory DEFAULT_INSTANCE = new MyHttpTransportFactory();

  private MyHttpTransportFactory() {}

  @Override
  public HttpTransport create() {
    return DEFAULT_TRANSPORT;
  }
}

static final class MyApplicationDefaultAuthCredentials extends AuthCredentials {

  private final GoogleCredentials googleCredentials;

  private static final ApplicationDefaultAuthCredentialsState STATE =
      new ApplicationDefaultAuthCredentialsState();

  private static class ApplicationDefaultAuthCredentialsState
      implements RestorableState<AuthCredentials>, Serializable {

    private static final long serialVersionUID = -8839085552021212257L;

    @Override
    public AuthCredentials restore() {
      try {
        HttpTransport transport = MyHttpTransportFactory.DEFAULT_INSTANCE.create();
        return new MyApplicationDefaultAuthCredentials(
            GoogleCredentials.getApplicationDefault(transport));
      } catch (IOException e) {
        throw new IllegalStateException(
            "Could not restore " + ApplicationDefaultAuthCredentials.class.getSimpleName(), e);
      }
    }

    @Override
    public int hashCode() {
      return getClass().getName().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
      return obj instanceof ApplicationDefaultAuthCredentialsState;
    }
  }

  public MyApplicationDefaultAuthCredentials(GoogleCredentials credentials) {
    googleCredentials = credentials;
  }

  public MyApplicationDefaultAuthCredentials() throws IOException {
    HttpTransport transport = MyHttpTransportFactory.DEFAULT_INSTANCE.create();
    googleCredentials = GoogleCredentials.getApplicationDefault(transport);
  }

  @Override
  public GoogleCredentials credentials() {
    return googleCredentials;
  }

  @Override
  public RestorableState<AuthCredentials> capture() {
    return STATE;
  }
}

DatastoreOptions options = DatastoreOptions.builder()
    .projectId(projectId)
    .namespace(NAMESPACE)
    .httpTransportFactory(MyHttpTransportFactory.DEFAULT_INSTANCE)
    .authCredentials(new MyApplicationDefaultAuthCredentials())
    .build();
```
and I see that credentials-related requests are correctly sent to the proxy. Can you share the exact requests that are not being sent to the proxy? Where is your app running and how are you setting default credentials (so that I can better replicate your issue)?

**Comment by mziccard** [comment]
This issue was fixed in #1375 where we dropped our AuthCredentials classes in favor of [google-auth-library-java](https://github.com/google/google-auth-library-java). Setting the `HttpTransportFactory` just became simpler:
```java
static final class MyHttpTransportFactory implements HttpTransportFactory {

  private static final HttpTransport DEFAULT_TRANSPORT = new NetHttpTransport.Builder()
      .setProxy(new Proxy(Proxy.Type.HTTP, new InetSocketAddress("localhost", 3128)))
      .build();
  static final MyHttpTransportFactory DEFAULT_INSTANCE = new MyHttpTransportFactory();

  private MyHttpTransportFactory() {}

  @Override
  public HttpTransport create() {
    return DEFAULT_TRANSPORT;
  }
}

DatastoreOptions options = DatastoreOptions.newBuilder()
    .setProjectId(PROJECT_ID)
    .setNamespace(NAMESPACE)
    .setHttpTransportFactory(MyHttpTransportFactory.DEFAULT_INSTANCE)
    .setCredentials(
        GoogleCredentials.getApplicationDefault(MyHttpTransportFactory.DEFAULT_INSTANCE))
    .build();
```
This will be available in the next release.
